docs/NEO-EDEN.md

```diff
@@ -0,0 +1,199 @@
+
+**But how do you debug that test?** Because you're running an interpreter, which runs bash commands via os.exec plus you execute golang program, which is written and compiled inside the repository. Debug printing would work, but I find debuggers much more useful and efficient most of the times. It is a matter of taste, but better to debugger at your disposal, than not to have it in this case. Even worse, how do you debug escript problem? You need to know it's internals, there's no Stackoverflow or Google by your side, nor there are books about it.
+
+**Bumping golang is actually manual action**: last but not least, we have to maintain custom test files which a basically copy-paste with one added function, bumping golang turns into some weird dances in that case.
+
+I believe that those reasons are good enough to switch for golang-only approach to write tests, which is called NeoEden (kudos to @shjala for the name)
+
+Also worst case what could happen with DSL is something like the Inner JSON effect [link](https://thedailywtf.com/articles/the-inner-json-effect)
+
+## Enter NeoEden. How tests look now?
+
+They look like [this](../tests/sec/sec_test.go) code listing is in the end of this section. Conceptually, Eden contains of two entities: an entity which allows you setup some kind of environment and deploy EVE there. Think Infrastructure as a Code, like Teraform, which allows me using golang code to deploy, this is openevec package; second entity is evetestkit, which serves as an abstraction over missing features of adam as controller such as keeping track of state and provide you with toolkit to test things, namely send commands and access EVE or its AppInstances to see if the certain file exists or certain request gives you certain response. Those two things let you create golang-only test like one below. You use openevec to setup environment and evetestkit to help you describe what is passing criterion for particular test. The common denominator between openevec and evetestkit is EdenSetupArgs, openevec is just a wrapper over that configuration, which defines your state. The goal is to be able to define your state, write custom functions to state, like this:
```

Comment:
Can you start with an overview (unless it is elsewhere, e.g. in another doc, and I missed it)? Something like:

In order to test eve with eden, there are a few steps you will need to take:

1. Start backends, controller, etc.
2. Read config
3. etc. etc.

Get the major ones.

Then describe what the golang functions are for those. 

Then go to an example. Or, you can put the example at the beginning, but explain that you will describe it step by step afterwards. In the example, it should be described step-by-step, maybe inline comments before each action describing what it does. Make it easy for someone reading it to understand.

Reply:
In terms of the content of this paragraph itself:

> Conceptually, Eden contains of two entities: an entity which allows you setup some kind of environment and deploy EVE there. Think Infrastructure as a Code, like Teraform, which allows me using golang code to deploy, this is openevec package; second entity is evetestkit, which serves as an abstraction over missing features of adam as controller such as keeping track of state and provide you with toolkit to test things, namely send commands and access EVE or its AppInstances to see if the certain file exists or certain request gives you certain response.

I wouldn't want to reference Terraform explicitly. Besides trademark issues, people will think this is much more than it is and might confuse them.

Maybe something like:

```
The eden golang test SDK for eve consists of two categories of functions:

1. `openevec` (link to the godoc for it): infrastructure manager. `openevec` enables you to deploy a controller, a backend for that controller, deploy an edge node, etc. (what else does it do?). It does not change anything on an individual edge node.
2. `evetestkit` (link to the godoc for it): provides fine-grained control over specific functions on individual edge nodes, like deploying applications or network instances, or deleting them. 

There is more you have, but it provides an entry structure that will make understanding the documented example easier.

Reply:
well, openevec allows you to deploy applications and to be fair, access node as well, I'd say evetestkit is a framework to easily describe what is the expected result for the test

Reply:
So my docs were off. General idea, though, would be really helpful.